File: nomdl/codegen/test/struct_recursive_test.go
package test

import (
	"testing"

	"github.com/attic-labs/noms/nomdl/codegen/test/gen"
	"github.com/stretchr/testify/assert"
)

func TestStructRecursiveChildren(t *testing.T) {
	assert := assert.New(t)

	root := gen.TreeDef{
		Children: []gen.TreeDef{
			gen.TreeDef{},
			gen.TreeDef{
				Children: []gen.TreeDef{
					gen.TreeDef{},
				},
			},
		},
	}.New()

	assert.Equal(uint64(2), root.Children().Len())
	assert.Equal(uint64(0), root.Children().Get(0).Children().Len())
	assert.Equal(uint64(1), root.Children().Get(1).Children().Len())
}

File: nomdl/codegen/test/struct-recursive-test.js
// @flow

import {assert} from 'chai';
import {suite, test} from 'mocha';
import {Tree, typeForTree, newListOfTree} from './gen/struct_recursive.noms.js';
import {newList, makeListType} from '@attic/noms';

suite('struct_recursive.noms', () => {
  test('constructor', async () => {
    const listOfTreeType = makeListType(typeForTree);
    const t: Tree = new Tree({children: await newList([
      new Tree({children: await newList([], listOfTreeType)}),
      new Tree({children: await newList([], listOfTreeType)}),
    ], listOfTreeType)});
    assert.equal(t.children.length, 2);

    const t2: Tree = new Tree({children: await newListOfTree([
      new Tree({children: await newListOfTree([])}),
      new Tree({children: await newListOfTree([])}),
    ])});
    assert.isTrue(t.equals(t2));
  });
});
